pkg/processor/build/util/decompress.go

```diff
@@ -37,31 +42,148 @@ func NewDecompressor(parentLogger logger.Logger) (*Decompressor, error) {
+
+	d.logger.DebugWith("Identified archive", "format", format, "input", input)
+
+	handler := func(ctx context.Context, file archiver.File) error {
+		d.logger.DebugWith("Extracting file", "sourcePath", file.Name, "targetPath", targetPath)
+
+		if filterFile, err := d.filterArchivedFile(file, targetPath); err != nil {
+			return errors.Wrap(err, "Failed to filter archived file")
+		} else if filterFile {
+			return nil
+		}
+
+		// copy the file to a new file with the same name in the target path
+		return d.extractArchivedFile(file, targetPath)
+	}
+
+	// want to extract something?
+	if extractor, ok := format.(archiver.Extractor); ok {
+		d.logger.DebugWith("Extracting archive", "sourcePath", sourcePath, "targetPath", targetPath, "extractor", extractor)
+		if err := extractor.Extract(ctx, input, nil, handler); err != nil {
+			return errors.Wrap(err, "Failed to extract archive")
+		}
+	}
+
+	return nil
+}
+
+func (d *Decompressor) extractArchivedFile(file archiver.File, targetPath string) error {
+	filePath := filepath.Join(targetPath, file.NameInArchive)
+
+	// create the directory if it doesn't exist
+	if err := os.MkdirAll(filepath.Dir(filePath), 0755); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed to create directory %s", filepath.Dir(filePath)))
+	}
+
+	// create the new file
+	outFile, err := os.Create(filePath)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed creating file :%s", filePath))
+	}
+	defer outFile.Close() // nolint: errcheck
+
+	// set the file mode
+	err = outFile.Chmod(file.Mode())
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed changing file mode for %s", filePath))
+	}
+
+	// open the file
+	input, err := file.Open()
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed opening file: %s", filePath))
+	}
+
+	// copy the file contents
+	_, err = io.Copy(outFile, input)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed writing file: %s", filePath))
+	}
+	return nil
+}
+
+func (d *Decompressor) filterArchivedFile(file archiver.File, targetPath string) (bool, error) {
+	// filter symlinks that point outside the archive or to an absolute path
+	if file.LinkTarget != "" && (strings.HasPrefix(file.LinkTarget, "..") || filepath.IsAbs(file.LinkTarget)) {
+		return true, nil
+	}
+
+	// check that the destination file path is not outside the target directory
+	fullTargetPath, err := filepath.Abs(path.Join(targetPath, file.NameInArchive))
+	if err != nil {
+		return true, errors.Wrap(err, fmt.Sprintf("Failed to get absolute path for %s", fullTargetPath))
+	}
+	if !strings.HasPrefix(fullTargetPath, targetPath) {
+		return true, nil
+	}
+
+	// filter files that their name is an absolute path
+	if filepath.IsAbs(file.NameInArchive) {
+		return true, nil
+	}
+
+	// filter files that their name contains ".."
+	if strings.HasPrefix(file.NameInArchive, "..") || strings.Contains(file.NameInArchive, "/..") {
+		return true, nil
+	}
+
+	// all good
+	return false, nil
+}
```

Comment:
Change this method to check the link targets and the file targets exactly how you did the file targets. (without the third name check)